src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -332,16 +332,30 @@
     (->> (map format-task info-from-jruby)
          (middleware-utils/json-response 200))))
 
+(defn task-file-uri-components
+  "The 'name' portion for a task implementation is a single component like
+  'foo.sh' and can never be nested in subdirectories. In that case we know the
+  'file' must be structured <environment>/<module root>/<module>/tasks/<filename>.
+  Other task files are the path relative to the module root, in the form
+  <module>/<mount>/<path> where <path> may have subdirectories. For those, we
+  just slice that off the end of the file path and take the last component that
+  remains as the module root."
+  [file-name file]
+  (if (str/includes? file-name "/")
+    (let [[module mount subpath] (str/split file-name #"/" 3)
+          module-root (fs/base-name (subs file 0 (- (count file) (count file-name))))]
```

Comment:
Curious why you would do "math" and string indexing instead of a regex (maybe something like `(clojure.string/replace file (re-pattern (str file-name "$")) "")` ?

Reply:
I think either way is probably fine. Since the original bug was caused by an imprecise use of regex, I was feeling somewhat disinclined to use that approach. It's actually somewhat tricky to formulate a regex that does _precisely_ 
what is intended here. For instance, your suggested code has a bug in that it doesn't _escape_ the string used to create the pattern, meaning special characters will retain their special meaning. _Unlikely_ to cause an issue, but certainly not impossible. Taking a substring sidesteps any of that and does _exactly_ what is meant: remove this 
substring from the end of the string.

A better way might be `(fs/base-name (subs file 0 (str/last-index-of file file-name)))` to avoid the finnicky math

Reply:
I think using `last-index-of` makes the intention clearer. 
I agree that avoiding a regex is ideal.

Reply:
Updated to use `last-index-of`. I think I couldn't wrap my brain around the logic on friday :D

Reply:
TIL `last-index-of`. Clojure is so cool.